[addons] binary cache update is not triggered on enable/disable of addons

xhaggi opened the pull request:
Enabling/disabling of VFS addons does not trigger update of binary addon caches. This PR will fix the issue.

@mkortstiege mind taking a look.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->
Runtime tested on MacOS

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Improvement (non-breaking change which improves existing functionality)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

xhaggi committed:
[addons][fix] binary cache update is not triggered on enable/disable of addons

AlwinEsch reviewed (commented):
Can be maybe complicated if e.g. VFS addon do something.

The ReInstalled and UnInstalled becomes called after the addon is changed or deleted.

Here where I change for PVR:
https://github.com/AlwinEsch/kodi/pull/35/commits/82b0d7aabbf629debaa2c539deae629d9e208296

xhaggi commented:
Not sure what you mean with do something. IMO if an addon becomes disabled or enabled the cache should be updated to reflect the current state. ATM if you enabled the installed vfs.rar addon you need to restart Kodi to use the addon. That's not very nice and fixed by this changes.

AlwinEsch commented:
Thats complete true, I mean if a vfs addon e.g. unzip is unpacking a zip file and becomes deinstalled. The CVFSAddonCache becomes informed after the addon is deleted.

xhaggi commented:
Ok let me know what I have to change and I'll do that.

AlwinEsch commented:
Have looked, I think you can use your way (a VFS addon is not always running) and implement also something by use of https://github.com/xbmc/xbmc/blob/master/xbmc/addons/Addon.h#L57. To check before the addon is changed.

Have currently not a direct idea how this `IsInUse` can be used from VFS, maybe you?

xhaggi commented:
Not really. We could go with this and take an eye on it. Let me know if we could merge it in.

AlwinEsch commented:
Ok from my side

AlwinEsch commented on xbmc/addons/VFSEntry.cpp:
```diff
@@ -70,8 +70,13 @@ VFSEntryPtr CVFSAddonCache::GetAddonInstance(const std::string& strId, TYPE type
+      typeid(event) == typeid(AddonEvents::ReInstalled) ||
+      typeid(event) == typeid(AddonEvents::UnInstalled))
+  {
     Update();
```
Only maybe add a:
```cpp
  ADDON::AddonPtr addon;
  if (CServiceBroker::GetAddonMgr().GetAddon(enableEvent->id, addon, ADDON_VFS, false))
    Update();
```
to prevent not wanted update if e.g. YouTube addon is changed.

xhaggi replied:
Sure will add it.

xhaggi replied:
This won't work for event `UnInstalled` because `GetAddon()` will always return false. We need to update the cache in that case.

xhaggi committed:
[addons] add convenient method to determine if addon is of supplied type

xhaggi committed:
[addons] refactor AddonEvent structs

* move id from derived structs to struct AddonEvent
* drop deprecated addon event InstalledChanged

xhaggi commented:
@AlwinEsch pushed some improvements after our discussion.

The first commit move the addon id to `AddonEvent` struct so that we don't not need to cast to get the addon id out of an event. It also drops the special `InstalledChanged` event in favor of a new event `Initilized` which is only triggered in `FindAddons`. This will reduce the events triggered in case of `UnInstalled` + `ReInstalled` where you also publish the `InstalledChanged` before. I've adjust all places where `InstalledChanged` was used.

The second commit adds a convenient method to check if an addon has a given type.

Hope this will make things a bit easier.

AlwinEsch commented:
From my side perfect, to have addon id defined on the base structure makes it much more easy to identify the addon type.

Thanks a lot!

xhaggi commented:
> From my side perfect, to have addon id defined on the base structure makes it much more easy to identify the addon type.

Cool. Force-pushed everything. Mind take a look again and finally sign-off ;)

AlwinEsch commented on xbmc/addons/AddonEvents.h:
```diff
@@ -25,27 +25,34 @@ namespace ADDON
+    /**
+     * Emitted after initilizing add-ons.
+     */
+    struct Initilized : AddonEvent
```
A small typo should be "Initialized"

xhaggi replied:
Oh thx, will address it.

AlwinEsch commented on xbmc/addons/AddonManager.cpp:
```diff
@@ -654,6 +654,12 @@ bool CAddonMgr::GetAddon(const std::string &str, AddonPtr &addon, const TYPE &ty
+  return GetAddon(id, addon, type, false);
+}
+
 bool CAddonMgr::FindAddons()
```
The FindAddons is also called on `bool CAddonMgr::LoadAddon(const std::string& addonId)`. Maybe add a `bool CAddonMgr::FindAddons(bool initialFind)` or prevent the `m_events.Publish(AddonEvents::Initilized());` from this function complete.

xhaggi replied:
Then we should move `m_events.Publish(AddonEvents::Initilized());` to `Init`

AlwinEsch commented on xbmc/addons/VFSEntry.cpp:
```diff
@@ -70,8 +70,18 @@ VFSEntryPtr CVFSAddonCache::GetAddonInstance(const std::string& strId, TYPE type
 {
-  if (typeid(event) == typeid(AddonEvents::InstalledChanged))
+  if (typeid(event) == typeid(AddonEvents::Initilized) ||
+      typeid(event) == typeid(AddonEvents::UnInstalled))
```
The `AddonEvents::UnInstalled` also to the check below.

xhaggi replied:
As I said `UnInstalled` can't be used in conjunction with `HasType` because the addon is gone

AlwinEsch commented on xbmc/peripherals/bus/virtual/PeripheralBusAddon.cpp:
```diff
@@ -332,9 +332,11 @@ void CPeripheralBusAddon::GetDirectory(const std::string &strPath, CFileItemList
-      typeid(event) == typeid(ADDON::AddonEvents::InstalledChanged))
+      typeid(event) == typeid(ADDON::AddonEvents::ReInstalled) ||
+      typeid(event) == typeid(ADDON::AddonEvents::UnInstalled))
     UpdateAddons();
```
Add also a `if (CServiceBroker::GetAddonMgr().HasType(event.id, ADDON_PERIPHERALDLL))` check.

xhaggi replied:
Sure, will add the same here

AlwinEsch committed:
[addons] add new blocking events to CBinaryAddonManager

This are needed to have it informed before the related addons becomes
informed.

AlwinEsch committed:
[addons] change event handling on CPeripheralBusAddon

tamland commented:
>It also drops the special InstalledChanged event in favor of a new event Initilized which is only triggered in FindAddons. This will reduce the events triggered in case of UnInstalled + ReInstalled where you also publish InstalledChanged before. I've adjust all places where InstalledChanged was used.

Don't. It is deprecated for a reason: you do not need it. This event is just a old quick solution. If you listen to the new Enabled, ReInstalled, and UnInstalled, it's redundant.

xhaggi commented:
@tamland I noticed that and there was one place (`FindAddons`) where `InstalledChanged` is published that does not fit to any of the events `Enabled`, `ReInstalled` and `UnInstalled`. If we don't need an event to be published in that case I'll drop the `Initialized` event.

xhaggi commented:
@AlwinEsch merged your changes into the branch. Runtime tested and works for me too.
@tamland we dropped `InstalledChanged` now.

xhaggi commented:
jenkins build this please

tamland commented on xbmc/addons/AddonEvents.h:
```diff
@@ -54,34 +62,31 @@ namespace ADDON
     };
 
+    /**
+     * Emitted after the add-on has been unloaded.
```
This is wrong. Both Load and Unload are before (that's the entire point). Actually, they should not even be part of AddonEvent. That's just more stuff from #12473 that wasn't completed.

tamland commented:
Come on guys. Adding Load/Unload to the existing event handler together with the async ones is not how this was supposed to be used. I tried to demonstrate show how to it safely in https://github.com/xbmc/xbmc/commit/8bd7213d2ec081717f3c3f4fc0ed43eea843e019. Neither CPeripheralBusAddon, CVFSAddonCache, CBinaryAddonManager is 'fixed'. It now just contains race conditions instead, so TBC. when you notice binary addons randomly failing to be detected.

xhaggi commented:
@tamland ok will push a PR that removes the use of Load/Unload in CPeripheralBusAddon, CVFSAddonCache, CBinaryAddonManager
